src/votekit/models.py

```diff
@@ -1,95 +1,292 @@
 from .pref_profile import PreferenceProfile
-from .utils import recursively_fix_ties, fix_ties
+import pandas as pd
+from .utils import score_dict_to_tos_ranking
+from typing import Callable, Optional
 
 
 class Election(ABC):
     """
-    Abstract base class for election types. Includes functions to resolve input
-    ties included in PreferenceProfile.
+    Abstract base class for election types.
+
+    Parameters:
+        profile (PreferenceProfile): the initial profile of ballots.
+        score_function (Callable[[PreferenceProfile], dict[str, float]], optional): A function
+            that converts profiles to a score dictionary mapping candidates to their current score.
+            Used in creating ElectionState objects and sorting candidates in Round 0. If None, no
+            score dictionary is saved and all candidates are tied in Round 0. Defaults to None.
+        sort_high_low (bool, optional): How to sort candidates based on `score_function`. True sorts
+            from high to low. Defaults to True.
 
     Attributes:
-        profile: a PreferenceProfile.
-        ballot_ties: an optional Bool, defaults to True. If True, resolve ties on ballots.
+        election_states (list[ElectionState]): a list of election states, one for each round of
+            the election. The list is 0 indexed, so the initial state is stored at index 0, round 1
+            at 1, etc.
+        score_function (Callable[[PreferenceProfile], dict[str, float]], optional): A function
+            that converts profiles to a score dictionary mapping candidates to their current score.
+            Used in creating ElectionState objects. Defaults to None.
+        length (int): the number of rounds of the election.
     """
 
-    def __init__(self, profile: PreferenceProfile, ballot_ties: bool = True):
-        if ballot_ties:
-            self._profile = self.resolve_input_ties(profile)
-        else:
-            self._profile = profile
+    def __init__(
+        self,
+        profile: PreferenceProfile,
+        score_function: Optional[
+            Callable[[PreferenceProfile], dict[str, float]]
+        ] = None,
+        sort_high_low: bool = True,
+    ):
+        self._profile = profile
+        self.score_function = score_function
+        self.sort_high_low = sort_high_low
+        self.election_states: list[ElectionState] = []
+        self._run_election()
+        self.length = len(self.election_states) - 1
+
+    def get_profile(self, round_number: int = -1) -> PreferenceProfile:
+        """
+        Fetch the PreferenceProfile of the given round number.
+
+        Args:
+            round_number (int, optional): The round number. Defaults to -1,
+                which accesses the final profile.
+
+        Returns:
+            PreferenceProfile
+
+        """
+        if round_number < -1:
+            raise ValueError("round_number must be -1 or non-negative.")
+
+        if round_number == -1:
+            round_number = len(self)
```

Comment:
This likely doesn't cause an issue now since an implementation of `_run_step` should do nothing if the election is over, but let's not rely on that always being the case and be intentional with our indexing so that we don't accidentally end up with undefined behaviour

Reply:
This should be okay actually; the len(self) in a 3 round election is 3, so the function runs 3 times.